### ISSUE TITLE
Summarize skipped jobs at the end of an SoS run

### ISSUE THREAD

**Issue by gaow** [vatlab/SoS#925]
A suggestion from a colleague of mine: it is of interest to know at the end of an SoS run how many jobs has been skipped vs actually executed in repeated executions. He proposed interface like:

```
INFO: {N} jobs completed ({X} ignored due to previous runs, {Y} re-executed due to signature mismatch and {N - X - Y} newly executed).
```

It seems a reasonable request. Can we communicate back this information and store them to `env` for output it at the end?

**Comment by BoPeng** [comment]
But what is the definition of job? We have `action` level, `task` level, and `step` level signatures.

**Comment by gaow** [comment]
I'd say the smallest unit of execution, at `action` level, ie, whatever executed via commandline `exec scipt.file`?

**Comment by gaow** [comment]
^ Bump this one, if not too complicated :)

**Comment by BoPeng** [comment]
Something like

> INFO: Workflow default (ID=96b2069d1991e995) is executed successfully with 1 completed step, 5 completed input groups, 1 skipped step.

You can adjust how these are outputted [here](https://github.com/vatlab/SoS/blob/master/src/sos/workflow_executor.py#L785)

**Comment by gaow** [comment]
Great, thanks a lot! It feels more secure and in control now of what has been executed. I'll edit the output info text and push back here.

**Comment by BoPeng** [comment]
OK. Note that we are not counting failed tasks/steps here because they would terminate the workflow.

**Comment by gaow** [comment]
Okay I think I have 2 issues that goes beyond simply editing the text ... I hope they are not hard to address under the current framework:

1. For nested workflow, the information is a bit mixed:

```
[GW] sos run test.dsc 
INFO: Executing default_0: 
INFO: input:   []
INFO: Executing a_0: 
INFO: input:   []
INFO: a_0 (index=0) is ignored due to saved signature
INFO: output:   [file_target('1.txt')]
INFO: Workflow a (ID=fcf71e8f520500e2) is executed successfully with 1 skipped step.
INFO: output:   []
INFO: Workflow default (ID=c619da916ea6a159) is executed successfully with 1 completed step.
```

I'd expect the last line be `1 skipped step` as well? Or, `1 skipped sub-workflow`? the MWE is:

```
[a]
output: "1.txt"
run:
  touch 1.txt

[default]
sos_run('a')
```

2. I am not sure why the status of `input` are reported not `output`, eg `6 skipped inputs`. It seems more natural to report skipped outputs, because we'll know that those output files were not re-generated, right?

**Comment by BoPeng** [comment]
1. I will have a look at subworkflows, I think all subworkflows are stacked and counted together, which is not necessarily a bad thing.

2. That is `input groups`, not `input files`.

**Comment by gaow** [comment]
> That is input groups, not input files.

I see. That means those jobs associated with those input groups are not executed. But the terminology is confusing ... I'd suggest we replace `input groups` with just `jobs` or `tasks`?

**Comment by BoPeng** [comment]
Well, `job` is generally referred by us as `PBS jobs`, `task` is reserved for sos tasks.

**Comment by gaow** [comment]
True ... Does `command` sounds better?

**Comment by BoPeng** [comment]
No. I have tried to give a proper name to this thing and I used "step process" before but that name does not show the fact that each process uses a subset of input files. However, input group is not accurate either because we can do `for_each='i'` without dividing input files.

**Comment by BoPeng** [comment]
substep?

**Comment by gaow** [comment]
> that name does not show the fact that each process uses a subset of input files

I'm not sure why that matters to stress subset of input files. `process` or `instance` sounds appropriate to me. How about `step instance`?

**Comment by BoPeng** [comment]
`substep` means a `step` needs a number of `substep`s to complete and seems appropriate. It is short and easier to understand than `instance`.

**Comment by BoPeng** [comment]
Google substep does come up with some usages, and we have subworkflow without hyphen as well.

**Comment by gaow** [comment]
Indeed, so `substep` it is!

**Comment by gaow** [comment]
Great, i'll make minor adjustments after the nested workflow problem is resolved (is it fixable? see [my MWE](https://github.com/vatlab/SoS/issues/925#issuecomment-389936055))

**Comment by BoPeng** [comment]
Does it needs to be fixed? Do you want to aggregate steps etc by subworkflow, or report by each subworkflow? Or you just want to get a count for subworkflow?

**Comment by gaow** [comment]
Whichever more convenient is fine, as long as there is no inconsistency in logic. Because every step has their own statistic anyways, which is what we look at when we dig into it. What I meant by logic inconsistency is this message in my MWE:

```
INFO: Workflow a (ID=fcf71e8f520500e2) is executed successfully with 1 skipped step.
INFO: output:   []
INFO: Workflow default (ID=c619da916ea6a159) is executed successfully with 1 completed step.
```

So this is because the nested workflow has just 1 step, and it is skipped, then I'd expect the aggregated behavior is also saying that the entire subworkflow (or whatever it is called, `step` is fine too) is skipped. Otherwise the message here is a bit confusing.

**Comment by BoPeng** [comment]
This patch aggregates steps and tasks from subworkflows but does not show the number of subworkflows. You can see commented code in the patch another possibility to show number of workflows (and does not aggregated steps within subworkflows). You can experiement with either.

**Comment by gaow** [comment]
Great! I think the commented out version (as is) is better. I changed the wording slightly.  

There is hopefully this one last corner case:

```
[a]
output: "1.txt"
run:
  touch 1.txt

[b]
output: "1.txt"
sos_run("a")

[default]
depends: sos_step('b')
output: "1.txt"
```

Then running it twice, I get:

```
INFO: Workflow default (ID=2913a222cbff8a93) is executed successfully with 1 completed step and 2 ignored steps.
```

That `1 completed step`, possibly referring to step `b`, is not accurate. Although changing `default` to using nested workflow `sos_run('b')` properly aggregates over nested workflows. Otherwise it looks great. I guess I am mostly interested in reading the last line of summary because this is what tells me exactly to what extend has my pipeline has changed (or not changed).

**Comment by gaow** [comment]
Right, sorry I meant to say the completed step possibly refers to the dependency on `b`, because with the following nested workflow:

```
[a]
output: "1.txt"
run:
  touch 1.txt

[b]
output: "1.txt"
sos_run("a")

[default]
output: "1.txt"
sos_run("b")
```

the end result is then 

```
INFO: Workflow default (ID=c7634f0d395ae5c4) is skipped with 1 ignored step.
```

which aggregates it properly.

**Comment by gaow** [comment]
Yes I figured, because I tried to fix it but cannot find where information for subworkflows are stored directly ... if fixing it makes the code more complicated in a way we do not like, then I suppose we can live with it ...

**Comment by BoPeng** [comment]
The logic was added to allow sos to dig into subworkflows to verify their signatures because users might change a subworkflow but not the wrapper step, and expect sos to detect the changes. This does cause problems like this and problems with re-execution of long-running scripts if they are in a step with subworkflow.
